src/build/_helpers.py

```diff
@@ -0,0 +1,117 @@
+import functools
+import os
+import subprocess
+import sys
+
+from typing import TYPE_CHECKING, AbstractSet, Callable, Iterator, Mapping, Optional, Sequence, Tuple, TypeVar, Union
+
+
+ConfigSettingsType = Mapping[str, Union[str, Sequence[str]]]
+PathType = Union[str, 'os.PathLike[str]']
+
+if TYPE_CHECKING:
+    from typing_extensions import Literal, Protocol
+
+    Distribution = Literal['sdist', 'wheel']
+    WheelDistribution = Literal['wheel']
+
+    class RunnerType(Protocol):
+        def __call__(
+            self, cmd: Sequence[str], cwd: Optional[PathType] = None, env: Optional[Mapping[str, str]] = None
+        ) -> None:
+            """
+            Run a command in a Python subprocess.
+
+            The parameters mirror those of ``subprocess.run``.
+
+            :param cmd: The command to execute
+            :param cwd: The working directory
+            :param env: Variables to be exported to the environment
+            """
+
+    class _Pep517CallbackType(Protocol):
+        def __call__(
+            self, cmd: Sequence[str], cwd: Optional[PathType] = None, extra_environ: Optional[Mapping[str, str]] = None
+        ) -> None:
+            ...
+
+
+def default_runner(cmd: Sequence[str], cwd: Optional[PathType] = None, env: Optional[Mapping[str, str]] = None) -> None:
+    subprocess.run(cmd, check=True, cwd=cwd, env=env)
+
+
+def quiet_runner(cmd: Sequence[str], cwd: Optional[PathType] = None, env: Optional[Mapping[str, str]] = None) -> None:
+    subprocess.run(cmd, check=True, cwd=cwd, env=env, stdout=subprocess.DEVNULL)
+
+
+def rewrap_runner_for_pep517_lib(
+    values: Union['RunnerType', Tuple['RunnerType', Optional[Mapping[str, str]]]]
+) -> '_Pep517CallbackType':
+    if isinstance(values, tuple):
+        runner, env = values
+    else:
+        runner = values
+        env = None
+
+    @functools.wraps(runner)
+    def inner(cmd: Sequence[str], cwd: Optional[PathType] = None, extra_environ: Optional[Mapping[str, str]] = None) -> None:
+        local_env = os.environ.copy() if env is None else dict(env)
+        if extra_environ:
+            local_env.update(extra_environ)
+
+        runner(cmd, cwd=cwd, env=local_env)
+
+    return inner
+
+
+def check_dependency(
```

Comment:
> I assume there's a reason you move this here, right? Do we have issues with circular imports?

Yes, `__init__` would import `env` and `env` import `check_dependency` and `PathType` from `__init__`.  This manifests as an actual import cycle when building the docs with `TYPE_CHECKING` being `True`.

Reply:
How were you importing? That should not happen when using absolute imports.

Reply:
I don't think absolute imports is a good way to avoid import cycles. They're still there, you're just imposing artificial limitations on your package in a manner which is obscure and probably implementation specific.